packages/contracts-bedrock/src/universal/OptimismMintableERC20FactoryInterop.sol

```diff
@@ -0,0 +1,100 @@
+    /// @param localToken  Address of the local token.
+    /// @param remoteToken Address of the remote token.
+    event DeploymentStored(address indexed localToken, address indexed remoteToken);
+
+    /// @notice Storage slot where hashOnion is stored at.
+    /// keccak256(abi.encode(uint256(keccak256("hashOnion")) - 1)) & ~bytes32(uint256(0xff));
+    bytes32 internal constant HASH_ONION_SLOT = 0xae706b81046b35e941cb9737b644ec489aa4690fffd2c3f501ccbed8fd5b0400;
+
+    /// @notice Initial layer on the hash onion.
+    /// keccak256(abi.encode(0));
+    bytes32 internal constant INITIAL_ONION_LAYER = 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563;
+
+    /// @notice Semantic version.
+    /// @custom:semver +interop
+    function version() public view override returns (string memory) {
+        return string.concat(super.version(), "+interop");
+    }
+
+    /// @notice hashOnion getter.
+    /// @return hashOnion_ The hashOnion value.
+    function hashOnion() public view returns (bytes32 hashOnion_) {
+        assembly {
+            hashOnion_ := sload(HASH_ONION_SLOT)
+        }
+    }
+
+    /// @notice Verifies the hashOnion and stores the provided addresses.
+    ///         Stores the deployments of the provided tokens if the computed hashOnion is valid. It can be considered
+    ///         complete when the hashOnion matches the initial layer.
+    /// @param _localTokens  Array of local token addresses.
+    /// @param _remoteTokens Array of remote token addresses.
+    /// @param _startingInnerLayer The starting inner layer of the hashOnion.
+    function verifyAndStore(
+        address[] calldata _localTokens,
+        address[] calldata _remoteTokens,
+        bytes32 _startingInnerLayer
+    )
+        external
+    {
+        bytes32 _hashOnion = hashOnion();
+        if (_hashOnion == 0) revert HashOnionNotSet();
+        if (_hashOnion == INITIAL_ONION_LAYER) revert OnionAlreadyPeeled();
+        if (_localTokens.length != _remoteTokens.length) revert TokensLengthMismatch();
+
+        // Unpeel the hash onion and store the deployments
+        bytes32 innerLayer = _startingInnerLayer;
+        for (uint256 i; i < _localTokens.length; i++) {
+            innerLayer = keccak256(abi.encodePacked(innerLayer, abi.encodePacked(_localTokens[i], _remoteTokens[i])));
+            deployments[_localTokens[i]] = _remoteTokens[i];
+            emit DeploymentStored(_localTokens[i], _remoteTokens[i]);
+        }
+
+        if (innerLayer != _hashOnion) revert InvalidHashOnion();
+        assembly {
+            sstore(HASH_ONION_SLOT, _startingInnerLayer)
+        }
+    }
+
+    /// @notice One-time setter for the hashOnion value to be called by the ProxyAdmin.
+    /// @param _hashOnion The new hashOnion value.
+    function setHashOnion(bytes32 _hashOnion) external {
+        if (msg.sender != Predeploys.PROXY_ADMIN) revert Unauthorized();
```

Comment:
I think this is wrong. Not sure about this one, but to my understanding PROXY_ADMIN has an owner, which is the gov address. That should be the authorized sender, not the PROXY_ADMIN contract. But pls, check this claim

Reply:
This is a good catch. I updated to the owner of `ProxyAdmin`.
But we're not still completely sure if that's the correct address.

So I'll leave a TODO to research more about this, we can raise the question once we push the spec so they have full context of this topic.